docs: how to add functions and modules (#1077)
AFM-SPM/TopoStats: docs/contributing/adding_modules.md

# How to add modules

## Basic Functionality

This should reside in `topostats/` or optionally `topostats/<sub_dir>`

Unit tests should be written for each function, classes and method and integration tests to ensure the overall

> functionality is as expected now and in the future should refactoring be undertaken

Reviewer: ```suggestion
functionality is as expected now and in the future should refactoring be undertaken.
```

    type=int,
    required=False,
    help="Maximum length of scars in pixels",
)
# Run the relevant function with the arguments
filter_parser.set_defaults(func=run_modules.filters)
```

1. Create a subparser with `<module_name>_parser = subparsers.add_parser(...)`.
2. Add arguments to subparser with `<module_name>_parser.add_argument(...)`, include...

- Single letter flag (optional).
- Longer flag (required).
- `dest` the destination to which the variable will be saved. This should match the corresponding value in the
  `default_config.yaml` which makes updating the configuration options straight-forward and will occur automatically
  (the code is in place you don't need to add anything, just ensure the names match).
- `type` should ideally be included as this helps define the type of the stored variable, particularly useful if for
  example paths are provided which should be stored as [`Path` pathlib][python_pathlib] objects.
- `default` a sensible default value, typically `None` though so that the value in `default_config.yaml` is used.
- `help` a useful description of what the configuration option changes along with possible options.

<!-- markdownlint-disable MD029 -->

> 3. Set the default function that will be called with `<module_name_parser.set_defaults(func=run_modules.<function>)`

Reviewer: ```suggestion
3. Set the default function that will be called with `<module_name>_parser.set_defaults(func=run_modules.<function>)`
```

#### Running in Parallel

One of the main advantages of TopoStats is the ability to batch process images. As modern computers typically have
multiple cores it means the processing can be done in parallel making the processing faster. To achieve this we use
wrapper functions in `topostats/run_modules.py` that leverage the [`Pool`][python_pool] multiprocessing class.

Your function should first load the `.topostats` files you wish to process that are found from the user specified path
(default being the current directory `./`).

You then need to define a `partial()` function with the `topostats/processing.py<your_module_processing_function>` as
the first argument and the remaining configuration options. These will typically be a subset/list from the
`default_config.yaml` where the configuration options use the _exact_ same names as the arguments of the function you
defined your `topostats/processing.py`. As mentioned above keeping configuration names consistent between configuration
files and functions means [`**kwargs`][python_kwargs] can be used when passing options to functions.

> Continuing with out example lets look at the [`topostats.processing.run_filters()`][topostats_entry_point_filters]

Reviewer: ```suggestion
Continuing with our example let's look at the [`topostats.processing.run_filters()`][topostats_entry_point_filters]
```

                        Images(
                            array, **plotting_config["plot_dict"][plot_name]
                        ).plot_histogram_and_save()
                    except AttributeError:
                        LOGGER.info(
                            f"[{filename}] Unable to generate plot : {plot_name}"
                        )
            plotting_config["run"] = True
        # Always want the 'z_threshed' plot (aka "Height Thresholded") but in the core_out_path
        plot_name = "z_threshed"
        plotting_config["plot_dict"][plot_name]["output_dir"] = core_out_path
        Images(
            filters.images["gaussian_filtered"],
            filename=filename,
            **plotting_config["plot_dict"][plot_name],
        ).plot_and_save()
        LOGGER.info(f"[{filename}] : Filters stage completed successfully.")
        return filters.images["gaussian_filtered"]
    # Otherwise, return None and warn that initial processing is disabled.
    LOGGER.error(
        "You have not included running the initial filter stage. This is required for all subsequent "
        "stages of processing. Please check your configuration file."
    )
    return None
```

> This instantiates (creates) and object `filters` of the class `Filters` with the supplied options and then runs the

Author: ```suggestion
This instantiates (creates) the object `filters` of the class `Filters` with the supplied options and then runs the
```
